circom/circom_test.go

```diff
@@ -0,0 +1,338 @@
+	Pub    sw_emulated.AffinePoint[T] `gnark:",public"`
+	RPub   sw_emulated.AffinePoint[T]
+
+	PlainChunks  []byte
+	Iv           [12]frontend.Variable `gnark:",public"`
+	ChunkIndex   frontend.Variable     `gnark:",public"`
+	CipherChunks []byte                `gnark:",public"`
+}
+
+func (c *MixEncryptionTest[T, S]) Define(api frontend.API) error {
+	MBytes := make([]frontend.Variable, len(c.PlainChunks))
+	for i := 0; i < len(c.PlainChunks); i++ {
+		MBytes[i] = c.PlainChunks[i]
+	}
+
+	CiphertextBytes := make([]frontend.Variable, len(c.CipherChunks))
+	for i := 0; i < len(c.CipherChunks); i++ {
+		CiphertextBytes[i] = c.CipherChunks[i]
+	}
+
+	cr, err := sw_emulated.New[T, S](api, sw_emulated.GetCurveParams[T]())
+	if err != nil {
+		return err
+	}
+	//check BigR=rG
+	cr.AssertIsOnCurve(&c.BigR)
+	api.Println("R check on curve ok")
+	BigR := cr.ScalarMulBase(&c.SmallR)
+	cr.AssertIsEqual(BigR, &c.BigR)
+	api.Println("R =rG check ok")
+	//check pub
+	cr.AssertIsOnCurve(&c.Pub)
+	api.Println("Pub check on curve ok")
+	//check RPub
+	cr.AssertIsOnCurve(&c.RPub)
+	api.Println("RPub check on curve ok")
+	RPub := cr.ScalarMul(&c.Pub, &c.SmallR)
+	cr.AssertIsEqual(RPub, &c.RPub)
+	api.Println("RPub =rPub check ok")
+	//generate key=hash(RPub)
+	uapi, err := uints.New[uints.U64](api)
+	nbBits := 8 * ((fp.Modulus().BitLen() + 7) / 8)
+	rawRpub := make([]uints.U8, 2*nbBits)
+	for i := range cr.MarshalG1(c.RPub) {
+		rawRpub[i] = uapi.ByteValueOf(cr.MarshalG1(c.RPub)[i])
+	}
+	hasher, err := zksha3.New256(api)
+	if err != nil {
+		return fmt.Errorf("hash function unknown ")
+	}
+	hasher.Write(rawRpub)
+	expected := hasher.Sum()
+	key := [32]frontend.Variable{}
+	for j := range key {
+		key[j] = expected[j]
+	}
+	api.Println("key generate ok")
+	//check aes process
+	aes := NewAES256(api)
+	gcm := NewGCM256(api, &aes)
+	gcm.Assert256(key, c.Iv, c.ChunkIndex, MBytes, CiphertextBytes)
+	api.Println("aes check ok")
+
+	return nil
+}
+
+func TestMixEncryption(t *testing.T) {
+	//check BigR=rG
+	_, g := secp256k1.Generators()
+	var r fr_secp.Element
+	_, _ = r.SetRandom()
+	SmallR := new(big.Int)
+	r.BigInt(SmallR)
+	var BigR secp256k1.G1Affine
+	BigR.ScalarMultiplication(&g, SmallR)
+	//generator PublicKey
+	source := rand.NewSource(time.Now().UnixNano())
+	rand := rand.New(source)
+	privKey, err := ecies.GenerateKey(rand, crypto.S256(), nil)
+	if err != nil {
+		return
+	}
+	var px fp.Element
+	px.SetInterface(privKey.PublicKey.X)
+	var py fp.Element
+	py.SetInterface(privKey.PublicKey.Y)
+	Pub := secp256k1.G1Affine{
+		px,
+		py,
+	}
+	//generator RPub=r*PublicKey
+	var RPub secp256k1.G1Affine
+	RPub.ScalarMultiplication(&Pub, SmallR)
+	RawRpub := RPub.RawBytes()
+	//generator key=hash(RPub)
+	key := RawRpub[:]
+	hasher := sha3.New256()
+	hasher.Write(key)
+	expected := hasher.Sum(nil)
+	keyBytes := [32]frontend.Variable{}
+	for i := 0; i < len(expected); i++ {
+		keyBytes[i] = expected[i]
+	}
+	//message
+	m := []byte{0x01, 0x02}
```

Comment:
Given that AES-GCM operates on 16-byte blocks, this leads to `numberBlocks = len(m) / 16 = 0` in the `Assert256`function.
As a result, maybe no constraints are imposed on the AES-GCM encryption process, effectively rendering the test ineffective

Reply:
maybe something like
```go
m := []byte{
    0x00, 0x01, 0x02, 0x03,
    0x04, 0x05, 0x06, 0x07,
    0x08, 0x09, 0x0A, 0x0B,
    0x0C, 0x0D, 0x0E, 0x0F,
}

numberBlocks := int(math.Ceil(float64(inputSize) / 16.0))

```

Reply:
Understand. I write these for convenience. Because the shared private key is actually a fixed-length ciphertext. I'll fix it